### ISSUE TITLE
Replace StoreResolver with StoreManager in RequestBuilder

### ISSUE THREAD

**Issue by shochdoerfer** [Smile-SA/elasticsuite#1010]
In the current implementation of the RequestBuilder the StoreResolver is used to resolve the store for the current request. The default implementation of the StoreResolver is using the information stored in the `store` cookie. In our case we use elasticsuite in a headless environment which basically means the cookie is not set and thus resolving the store fails.

I`d propose to use the StoreManager instead to get the current store as the StoreManager will respect the store view code found in the url. In case that is not wanted, the RequestBuilder should be changed to make it easier to overwrite the needed methods or introduce an interface to make it easier to provide a custom implementation.

### Preconditions

Magento Version : CE 2.2.5

ElasticSuite Version : 2.5.13

Environment : Developer & Production mode

Third party modules : -

**Comment by romainruaud** [comment]
I'm not sure why @afoucret did choose StoreResolver over StoreManager, but from what I can see in the code, StoreResolver does not only check in cookies : 

- it checks the `$_GET` parameter `__store` and fallbacks to cookies only if missing.
- it then fallbacks to the default store id.

We'll have a more detailed look on your issue but for now I can see several workaround : 

- rely on the `__store` parameter
- DI allow you to inject a custom implementation of `StoreResolverInterface` (where you would write a `getCurrentStoreId()` which fit your needs) inside the RequestBuilder (can be done by tweaking argument and/or using a virtual type)

Regards

**Comment by shochdoerfer** [comment]
Unfortunately the StoreResolver does not take the store view code into account that can be passed via the url. Having a seperate call in our headless application just because of this feels wrong ;)

I'd be happy if there is an RequestBuilder interface that I can implement. Currently I have to extend the default RequestBuilder and overwrite every single method which seems odd to me.

I'd happily contribute the needed changes, just let me know which direction you want me to take.

**Comment by romainruaud** [comment]
You do not have to overwrite our RequestBuilder.

You should just inject another StoreResolver into the request builder via the DI like this : 

```xml
    <type name="Smile\ElasticsuiteCore\Model\Search\RequestBuilder">
        <arguments>
            <argument name="storeResolver" xsi:type="object">Schochdoerfer\Mymodule\Model\StoreResolver</argument>
        </arguments>
    </type>
```

and the class should look like this : 

```php

namespace Schochdoerfer\Mymodule\Model;

class StoreResolver 
extends \Magento\Store\Model\StoreResolver
// or just implements \Magento\Store\Api\StoreResolverInterface
{
    public function getCurrentStoreId() 
    {
        //your custom logic. Maybe just inject StoreManager and return $storeManager->getCurrentStore().
    }
}
```

**Comment by shochdoerfer** [comment]
Good catch, thanks. Is this issue invalid then?

**Comment by romainruaud** [comment]
Hello,

imho Elasticsuite should work out of the box, even in a headless configuration, so we'll have to dig more into your issue. 

It may be possible that what you pointed here is a missing feature of Magento and not of Elasticsuite, you should probably propose the same issue on the Magento repository and explain that their legacy StoreResolver does not work properly on an headless environment. Maybe they'll be able to explain why it's the case and fix it, or give you more details about how you should be using it.

For now, I just gave you this quick workaround in order to : 

- help you fixing your problem quickly so this is not blocking your project anymore.
- ensure you will not override too much of Elasticsuite and will not be bothered when updating the module later.

So for me, issue is still valid and need a bit more investigation, but for now you should be fine with the workaround and able to go further.

Best regards

**Comment by shochdoerfer** [comment]
Just opened an issue in the Magento repo: https://github.com/magento/magento2/issues/16976

**Comment by shochdoerfer** [comment]
Apparently Magento thinks this is a feature request, not a bug report. So the "feature request" request can be found here: https://github.com/magento/community-features/issues/77

**Comment by afoucret** [comment]
PR #1052 have been merged and will be part of release 2.6.0 coming next week.

**Comment by shochdoerfer** [comment]
That came unexpected, cool. Thanks.
